lib/internal/test_runner/mock/mock.js

```diff
@@ -61,9 +61,23 @@ class MockFunctionContext {
   mockImplementationOnce(implementation, onCall) {
     validateFunction(implementation, 'implementation');
     const nextCall = this.#calls.length;
-    const call = onCall ?? nextCall;
-    validateInteger(call, 'onCall', nextCall);
-    this.#mocks.set(call, implementation);
+    if (onCall) {
+      validateInteger(onCall, 'onCall', nextCall);
+      this.#mocks.set(onCall, implementation);
+
+      return this;
+    }
+
+    const mocksSize = this.#mocks.size;
+    for (let i = 0; i < mocksSize; i++) {
+      if (!this.#mocks.has(i + nextCall)) {
+        this.#mocks.set(i + nextCall, implementation);
+        return this;
+      }
+    }
```

Comment:
I'm not sure if this is right either (although I haven't tested this variation). I think you want a counter, and you want to loop over all of the existing entries in `this.#mocks` to see if each entry should be included in the counter or not.

Reply:
This solution find the first empty implementation when you go incrementally, and then put it there

I don't see a problem with this and I would appreciate if you have a concrete example 😀

Reply:
This is a breaking change:

```js
test('foo', (t) => {
  let cnt = 0;
  function addOne() {
    cnt++;
    return cnt;
  }

  const fn = t.mock.fn(addOne);

  fn.mock.mockImplementationOnce(() => 999, 0);
  fn.mock.mockImplementationOnce(() => 888, 1);
  fn.mock.mockImplementationOnce(() => 777);

  console.log(fn());
  console.log(fn());
  console.log(fn());
});
```

This currently prints 777, 888, 1. With the changes in this PR, it now prints 999, 888, 777 (note that I'm not even using chaining here).

Reply:
I now this is a breaking change this is why I wrote that in the PR description

Reply:
Ah, I didn't notice that you had updated the PR description recently.

I'm not sure if chaining can be implemented without a breaking change. I personally prefer the simplicity of `mockImplementationOnce(() => 777)` always meaning "mock the next call" to now needing to mentally keep track of which calls are mocked in order to support chaining.

Reply:
@ErickWendel

Reply:
I think it can be implemented but I think it's gonna be unintuitive DX...

Reply:
> This is a breaking change:
> 
> ```js
> test('foo', (t) => {
>   let cnt = 0;
>   function addOne() {
>     cnt++;
>     return cnt;
>   }
> 
>   const fn = t.mock.fn(addOne);
> 
>   fn.mock.mockImplementationOnce(() => 999, 0);
>   fn.mock.mockImplementationOnce(() => 888, 1);
>   fn.mock.mockImplementationOnce(() => 777);
> 
>   console.log(fn());
>   console.log(fn());
>   console.log(fn());
> });
> ```
> 
> This currently prints 777, 888, 1. With the changes in this PR, it now prints 999, 888, 777 (note that I'm not even using chaining here).

I got that it's a breaking change but this behavior doesn't look right. 

I love the chaining possibility and I believe it's a great addition!

In the algorithm, shouldn't we just enqueue all `mockImpl` so we know the exact order?

Reply:
we do "enqueue" it by putting that in the next available location

Reply:
> we do "enqueue" it by putting that in the next available location

hmm, why is the order reversed on @cjihrig 's example? is it because of the second param?

Reply:
to show the breaking change